text/2023-04-26-color-outlines.md

```diff
@@ -0,0 +1,198 @@
+# API for color outlines
+
+In order to support [COLRv1] (and v0) fonts, the
+[skrifa](https://github.com/googlefonts/fontations/tree/main/skrifa)
+crate requires an API to expose a more sophisticated set of operations than
+are handled by the current [`Pen`](https://docs.rs/font-types/0.1.4/font_types/trait.Pen.html)
+trait that is used to generate basic outlines.
+
+## Proposal
+
+The color [specification][COLRv1] defines a directed acyclic graph of "paint"
+nodes with each describing an operation such as transform, clip or fill. An
+implementation is expected to traverse the graph in depth first order applying
+those operations in a sequence. While this is a reasonable representation for
+storage and analysis, exposing this data structure directly through a trait
+based callback interface is cumbersome.
+
+The proposed API would flatten the graph structure into a linear sequence of
+drawing commands, making use of stack semantics to capture the graph structure
+where necessary. The set of commands required to represent the graph is fairly
+small and maps nicely to most graphics APIs which already represent layering
+with a state stack.
+
+## Interface
+
+This is a rough outline of the proposed interface which would introduce a new
+`ColorPen` trait for accessing color outlines.
+
+### Supporting types
+
+Basic types that are used by the new `ColorPen` trait.
+
+```rust
+/// Extents of a rectangular region.
+pub struct BoundingBox;
+
+/// Affine transformation matrix.
+pub struct Transform;
+
+/// 32-bit RGBA unpremultiplied color.
+pub struct Color;
```

Comment:
If we don't do any gradient stop normalization or preparation for Skia specific shader behavior (which I find preferable due to differences), we don't need to interpolate any colors and we can decouple CPAL handling as well, and return color indices, and leave the rest to the client. I think this would be consistent with returning more values raw and as from the font. 

As a side note, [blackrenderer backends](https://github.com/BlackFoundryCom/black-renderer/tree/main/Lib/blackrenderer/backends) may be a good reference, but beware of gradient bugs in it, at least sweep is still broken (https://github.com/BlackFoundryCom/black-renderer/issues/17).

Reply:
Returning palette indices is definitely an option. If we do so, I would like skrifa to have a convenience method for resolving them.